compiler/src/org.graalvm.compiler.lir/src/org/graalvm/compiler/lir/hashing/Hasher.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2011, 2018, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.graalvm.compiler.lir.hashing;
+
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeSet;
+
+import org.graalvm.compiler.lir.gen.ArithmeticLIRGeneratorTool;
+
+import jdk.vm.ci.meta.JavaConstant;
+import jdk.vm.ci.meta.Value;
+
+public final class Hasher {
+
+    public static Optional<Hasher> forKeys(JavaConstant[] keys, double minDensity) {
+        if (keys.length <= 2) {
+            return Optional.empty();
+        } else {
+            int maxCardinality = (int) Math.round(keys.length / minDensity);
+            assertSorted(keys);
+            TreeSet<Hasher> candidates = new TreeSet<>(new Comparator<Hasher>() {
+                @Override
+                public int compare(Hasher o1, Hasher o2) {
+                    int d = o1.cardinality - o2.cardinality;
+                    if (d != 0) {
+                        return d;
+                    } else {
+                        return o1.effort() - o2.effort();
+                    }
+                }
+            });
+            long min = keys[0].asLong();
+            for (HashFunction f : HashFunction.instances()) {
+                for (int cardinality = keys.length; cardinality < maxCardinality; cardinality++) {
+                    if (isValid(keys, min, f, cardinality)) {
+                        candidates.add(new Hasher(f, cardinality, min));
+                        break;
+                    }
+                }
+            }
+            if (candidates.isEmpty()) {
+                return Optional.empty();
```

Comment:
Any idea how often this happens in practice?

Reply:
It happens often since `minDensity` is normally too low and the hasher stops the evaluation based on it. It's necessary to review the heuristic that makes the decision between the density of the table switch x the cost of other strategies. I'll remove the bonus parameter for now and follow up with a separate PR.